### DISCUSSION TITLE
Freezing layers YOLOv8 for fine-tuning

### DISCUSSION THREAD

**Discussion by Rayen023** [ultralytics/ultralytics#3862] [Q&A]
Hello, I'm currently attempting to freeze the backbone of YOLOv8 for fine-tuning purposes. As a baseline, I tried training the model with all layers frozen. Surprisingly, it produced favorable results. Initially, I believed this to be an issue with PyTorch's implementation of `requires_grad = False`. However, after perusing the documentation at the following link:

[Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers/#accuracy-comparison)

It appears this is the expected behavior. 
Consequently, I'm trying to understand how the weights and biases can still update even when `requires_grad = False` has been applied to all layers.

Here's my code:

```python
model = YOLO("yolov8l.yaml")  # Build a new model from scratch
for k, v in model.model.named_parameters(): 
    v.requires_grad = False  # Freeze all layers 

# Use the model
model.train(data=data_path, epochs=50, batch=32, imgsz=544, device=0)
```

I also tried to set the momentum and weight_decay to zero after freezing the layers, as one user suggested that the model might still be learning because these parameters might still provide access to gradients. Any insights would be appreciated.

**Comment by atharvavaidya14** [comment]
Hello,
Did you find any solution to the problem? I am also working with fine-tuning yolov8 with freezing the backbone. But if it is not getting frozen in the way they have mentioned in the documentation, then that's a big problem.

**Reply by glenn-jocher** [reply]
@atharvavaidya14 hello! When you're fine-tuning YOLOv8 and you want to freeze certain layers, setting `requires_grad = False` should indeed prevent those layers from updating during training. If you're observing changes in performance, it could be due to other parts of the model or training process that are still active.

It's important to ensure that all layers you intend to freeze are correctly set. Additionally, if you're using pre-trained weights, make sure to load them before freezing the layers. After freezing, no updates should occur to the weights and biases of those layers during backpropagation.

If you're still seeing unexpected behavior, it might be worth double-checking your training loop and optimizer setup to ensure that the frozen state of the parameters is respected throughout the training process.

Remember that fine-tuning typically involves freezing the early layers (like the backbone) and only training the later layers. If all layers are frozen, the model should not learn or change its predictions, as no learning is taking place.

If you continue to face issues, please provide more details about your training setup and any observations you've made, and we can look into it further. The YOLOv8 documentation on our modes and tasks might also offer additional insights into the training process. 📘🔍

**Comment by ksooklall** [comment]
From the [documentation](https://docs.ultralytics.com/modes/train/#arguments) you can supply `freeze` as a parameter to train. 
freeze, None, (int or list, optional) freeze first n layers, or freeze list of layer indices during training

**Reply by atharvavaidya14** [reply]
@ksooklall unfortunately, it does not work with yolov8-pose models. or maybe V8 in general. I haven't tried for other models
It gives this error- 
`SyntaxError: 'freeze' is not a valid YOLO argument`

with this code:
```
model = YOLO("yolov8x-pose.pt")
model.train(
        data=f'data/data.yaml',
        rect=True, 
        epochs=no_epochs, 
        patience=patience,
        imgsz=640, 
        batch=16, 
        close_mosaic=1, 
        save=True, 
        project=dir, 
        name=proj_name,
        warmup_epochs=0,
        freeze=2,
    )
```

**Comment by jamesquags** [comment]
how many layers is "the backbone"?

**Reply by pderrenger** [reply]
@jamesquags hello! In YOLOv8, the backbone typically refers to the initial set of layers responsible for feature extraction. The exact number of layers in the backbone can vary depending on the specific YOLOv8 model variant you're using (e.g., YOLOv8n, YOLOv8s, YOLOv8m, YOLOv8l, etc.).

For instance, in a YOLOv8n model, the backbone might consist of the first several layers up until the point where the feature maps are passed to the subsequent neck and head layers for further processing and detection.

If you're looking to freeze the backbone for fine-tuning, you would typically freeze these initial layers while leaving the rest of the network (neck and head) trainable. This allows you to leverage the pre-trained feature extraction capabilities while adapting the latter parts of the network to your specific dataset. 😊🔒

Here's a quick example of how you might freeze the backbone layers:

```python
model = YOLO('yolov8n.yaml')  # Load the model configuration
backbone_layers = model.model.backbone.named_parameters()  # Get backbone parameters
for name, param in backbone_layers:
    param.requires_grad = False  # Freeze the backbone layers
```

Keep in mind that the actual attribute names (`model.model.backbone`) may differ based on the model's implementation, so you might need to adjust the code accordingly. Happy fine-tuning!

**Reply by vgarg11** [reply]
This code doesnt work

**Reply by pderrenger** [reply]
@vgarg11 hi there! 😊 Could you please provide a bit more detail on what issue you're encountering with the code? Sharing the error message or describing the behavior you're observing would be super helpful. Also, if you could include the specific part of the code you're having trouble with, that would be great. This way, we can offer more targeted assistance. Thanks!

**Reply by manansaipi** [reply]
if i working with yolov8-seg, and i want t otransfer learning my custom dataset to the pretrain model(ylolv8-seg.pt), 
i tried this it does not work. 
yolo detect train data=transfer.yaml model=yolov8n-seg.pt epochs=100 imgsz=640 batch=-1 plots=True freeze=10
, after i successfully train using that command, the model cannot even detect a person, cats, etc (basic classess that the pretrained have)

here is my transfer.yaml

test: C:\Users\boben\obj detection project\test-transfer\test\images
train: C:\Users\boben\obj detection project\test-transfer\train\images
val: C:\Users\boben\obj detection project\test-transfer\valid\images

nc: 81  # Total number of classes including both original and COCO classes
names:
- 'person'
- 'bicycle'
- 'car'
- 'motorcycle'
- 'airplane'
- 'bus'
- 'train'
- 'truck'
- 'boat'
- 'traffic light'
- 'fire hydrant'
- 'stop sign'
- 'parking meter'
- 'bench'
- 'bird'
- 'cat'
- 'dog'
- 'horse'
- 'sheep'
- 'cow'
- 'elephant'
- 'bear'
- 'zebra'
- 'giraffe'
- 'backpack'
- 'umbrella'
- 'handbag'
- 'tie'
- 'suitcase'
- 'frisbee'
- 'skis'
- 'snowboard'
- 'sports ball'
- 'kite'
- 'baseball bat'
- 'baseball glove'
- 'skateboard'
- 'surfboard'
- 'tennis racket'
- 'bottle'
- 'wine glass'
- 'cup'
- 'fork'
- 'knife'
- 'spoon'
- 'bowl'
- 'banana'
- 'apple'
- 'sandwich'
- 'orange'
- 'broccoli'
- 'carrot'
- 'hot dog'
- 'pizza'
- 'donut'
- 'cake'
- 'chair'
- 'couch'
- 'potted plant'
- 'bed'
- 'dining table'
- 'toilet'
- 'tv'
- 'laptop'
- 'mouse'
- 'remote'
- 'keyboard'
- 'cell phone'
- 'microwave'
- 'oven'
- 'toaster'
- 'sink'
- 'refrigerator'
- 'book'
- 'clock'
- 'vase'
- 'scissors'
- 'teddy bear'
- 'hair drier'
- 'toothbrush'
- 'bollard'

roboflow:
  workspace: capstone-rbdau
  project: footpath-segmentation
  version: 4
  license: CC BY 4.0
  url: https://universe.roboflow.com/capstone-rbdau/footpath-segmentation/dataset/4

**Reply by pderrenger** [reply]
Hello there! 👋 It seems like you're experiencing issues with transfer learning on a custom dataset using YOLOv8-seg. From the description of your problem, it looks like the model is failing to detect basic classes after training, even though it was trained with both the original and new classes.

Here are a couple of things to check and try:

1. **Dataset Quality and Size**: Ensure that your custom dataset is properly labeled and is sufficiently large. Small or incorrectly labeled datasets can lead to poor training results.

2. **Class Overlap**: Are the classes in your custom dataset overlapping with COCO classes? If there's a mismatch in class IDs or if your dataset doesn't include all the original COCO classes along with the new ones, it may cause issues. Since you mentioned using `nc: 81`, ensure that the class IDs match exactly with those of the original model for the COCO classes.

3. **Freezing Layers**: Try adjusting the number of layers you're freezing. Freezing the first 10 layers might not be optimal for your specific dataset. Experiment with freezing fewer layers or even training the model without freezing any layers to see how it affects detection performance.

Here's a corrected command just in case:

```bash
yolo detect train data=transfer.yaml model=yolov8n-seg.pt epochs=100 imgsz=640 batch=-1 plots=True
```

And double-check your `transfer.yaml` to ensure it aligns closely with the format and class IDs originally trained with YOLOv8-seg.

Feel free to share any updates or further issues you encounter! 😊

**Reply by manansaipi** [reply]
Hi,

Thank you for your response.

Since I only want to train one class ('bollard' 81), my dataset (train, valid, test) only contains the bollard class (index 80). Do I need to include all basic classes (80 basic classes, from index 0 to 79) that the pre-trained model has into the (train, valid, test) datasets? If so, how many images do I need to include in the (train, valid, test) datasets?

I thought that when I want to do transfer learning, I don't need to include the images that the pre-trained model already has.

**Reply by pderrenger** [reply]
@manansaipi hey there! 😊

Absolutely, when doing transfer learning with a single class ('bollard' in your case), you don't need to include images for all the 80 basic classes the pre-trained model was trained on. You just need to focus on your specific class of interest for your train, valid, and test datasets.

The key step here is ensuring that your `transfer.yaml` is configured correctly for your single class. Since you're training on 'bollards' only, your `nc` (number of classes) should be set to 1, and your class names list should only include 'bollard'. Here's a small example:

```yaml
nc: 1
names: ['bollard']
```

As for the dataset size, there's no strict rule, but more data generally helps the model learn better. Aim for at least a few hundred images if possible, distributed across train, valid, and test sets to achieve noticeable results.

Happy training, and feel free to reach out if you have more questions! 🚀

**Reply by manansaipi** [reply]
Thankyou for your response @pderrenger :)

I a little bit confused now, based on https://github.com/orgs/ultralytics/discussions/1431 and https://github.com/ultralytics/ultralytics/issues/8876 I see @glenn-jocher said in order to transfer learning, need to include all basic classes on our data.yaml. which one is the correct one? i tried everything it does not work, here is what I tried

cli: 
yolo detect train data=train-seg.yaml model=yolov8n-seg.pt epochs=200 imgsz=640 batch=-1
and the train-seg.yaml
look like 

test: C:\Users\boben\obj detection project\combine_with_algi_dataset\test\images
train: C:\Users\boben\obj detection project\combine_with_algi_dataset\train\images
val: C:\Users\boben\obj detection project\combine_with_algi_dataset\valid\images

nc: 5
names: ['bollard', 'footpath', 'go', 'stop', 'tiang'], and my anotet class is is from 0 - 4

this only detect that 5 object and it wont detect other basic classes, and when I try to include all the classes

cli:
yolo detect train data=train-seg.yaml model=yolov8n-seg.pt epochs=200 imgsz=640  batch=-1
or
yolo detect train data=train-seg.yaml model=yolov8n-seg.pt pretrained=yolov8n.pt epochs=200 imgsz=640 batch=-1

here is the .yaml file lookalike 
nc: 81
names: ['person', '..', '..', 'bollard'], and my dataset only include bollard (bollard class is 80 ), it only detect bollard, not detect the other classes. what do I need to do :(

**Reply by manansaipi** [reply]
hi @pderrenger , 
so i have to include the all other class in order to transfer learning ?

**Reply by pderrenger** [reply]
Hello @manansaipi! 😊

Yes, for effective transfer learning with YOLOv8, especially when using a pre-trained model like `yolov8n-seg.pt`, it's recommended to include all the original classes in your dataset. This approach helps maintain the learned features for those classes, even if your primary focus is on new or specific classes like 'bollard'.

Here's a quick snippet for setting up your `.yaml` file:

```yaml
nc: 81  # Number of classes
names: ['person', 'bicycle', ..., 'bollard']  # Ensure 'bollard' is correctly indexed if it's a new class
```

Make sure your annotations for new classes are correctly indexed and included in your training data. This setup should help in retaining the detection capabilities for both the original and new classes.

Happy training! 🚀

**Reply by zohaibjan** [reply]
Can you freeze the backbone if you are using a yolov9.pt existing pretrained weights ?. When I tried the code but with the pt file it gave me this error: "AttributeError: 'DetectionModel' object has no attribute 'backbone'"

**Reply by pderrenger** [reply]
@zohaibjan hello!

For freezing the backbone in YOLOv9 models, it's important to ensure that the model architecture explicitly defines a 'backbone' attribute that can be accessed as you've tried. It seems like the YOLOv9 model structure might not have a directly accessible 'backbone' attribute like some other models.

A good approach would be to review the specific architecture of the YOLOv9 model you are using to identify how the layers are structured and named. You can then freeze the appropriate layers by their names or indices directly.

If you need further assistance on how to navigate the model's architecture or how to freeze specific layers, feel free to ask!

Happy coding! 🚀

**Reply by Sakethnoothi775** [reply]
> @zohaibjan hello!
> 
> For freezing the backbone in YOLOv9 models, it's important to ensure that the model architecture explicitly defines a 'backbone' attribute that can be accessed as you've tried. It seems like the YOLOv9 model structure might not have a directly accessible 'backbone' attribute like some other models.
> 
> A good approach would be to review the specific architecture of the YOLOv9 model you are using to identify how the layers are structured and named. You can then freeze the appropriate layers by their names or indices directly.
> 
> If you need further assistance on how to navigate the model's architecture or how to freeze specific layers, feel free to ask!
> 
> Happy coding! 🚀



Hello, i am trying to train a yolov9s for 2500 phone and 3000 book images i.e phone,book classes,  can you please suggest the exact number for parameter 'freeze' in training

**Reply by rishith2** [reply]
Did anyone figure it out, how to do transfer learning. Do we have to freeze the backbone(first 53 layers) and give dataset with base images and new class images? Or just pass the dataset(base+new) without freezing. 
I tried this:
model = YOLO("yolov8l.yaml")  
for k, v in model.model.named_parameters(): 
    v.requires_grad = False  # Freeze all layers 
model.train(data=data_path, epochs=50, batch=32, imgsz=544, device=0)

with dataset that only has new class images and freezing all layers,  it gives predictions for just new classes but not the base coco classes. Logically if youre freeezing all the layers, it shouldnt learn anything. Do we have to give the coco classes images as well? Does transfer learning happen without freezing, as i am giving the base classes images ?

**Reply by glenn-jocher** [reply]
For effective transfer learning with YOLOv8, you don't need to freeze the backbone. Simply include both the base and new class images in your dataset and train the model. Freezing layers is optional and typically used to retain specific learned features. Ensure your dataset and configuration file correctly reflect all classes. If issues persist, please verify with the latest package version.

**Comment by Y-T-G** [comment]
It's still getting updated because the `BatchNorm` layers are in `train` mode.

I freeze the layers completely [here](https://y-t-g.github.io/tutorials/yolov8n-add-classes/) by putting `BatchNorm` layers to `eval` mode during training.

**Reply by WindFreeFoliage** [reply]
So cool, Thanks

**Reply by glenn-jocher** [reply]
@WindFreeFoliage thank you for your kind words! The credit goes to the YOLO community and the Ultralytics team for their efforts. Let us know if you have any questions!
